include/oneapi/dpl/pstl/hetero/memory_ranges_impl_hetero.h

```diff
@@ -0,0 +1,237 @@
+            oneapi::dpl::__ranges::views::all_write(std::ranges::take_view(std::forward<_OutRange>(__out_r), __n_min)));
+    }
+
+    return {std::ranges::begin(__in_r) + __n_min, std::ranges::begin(__out_r) + __n_min};
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// __pattern_uninitialized_fill
+//---------------------------------------------------------------------------------------------------------------------
+
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R, typename _T>
+std::ranges::borrowed_iterator_t<_R>
+__pattern_uninitialized_fill(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r, const _T& __value)
+{
+    using _ValueType = std::ranges::range_value_t<_R>;
+
+    const auto __first = std::ranges::begin(__r);
+    const auto __last = __first + std::ranges::size(__r);
+
+    if constexpr (std::is_trivially_constructible_v<_ValueType, _T> &&     // required operation is trivial
+                  std::is_trivially_default_constructible_v<_ValueType> && // actual operations are trivial
+                  std::is_trivially_copy_assignable_v<_ValueType>)
+    {
+        oneapi::dpl::__internal::__ranges::__pattern_walk_n(
+            __tag, std::forward<_ExecutionPolicy>(__exec),
+            oneapi::dpl::__internal::__brick_fill<__hetero_tag<_BackendTag>, _ValueType>{_ValueType(__value)},
+            oneapi::dpl::__ranges::views::all_write(std::forward<_R>(__r)));
+    }
+    else
+    {
+        oneapi::dpl::__internal::__ranges::__pattern_walk_n(
+            __tag, std::forward<_ExecutionPolicy>(__exec),
+            oneapi::dpl::__internal::__op_uninitialized_fill<_T, std::decay_t<_ExecutionPolicy>>{__value},
+            oneapi::dpl::__ranges::views::all_write(std::forward<_R>(__r)));
+    }
+
+    return std::ranges::borrowed_iterator_t<_R>{__last};
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// destroy
+//---------------------------------------------------------------------------------------------------------------------
+
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R>
+std::ranges::borrowed_iterator_t<_R>
+__pattern_destroy(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r)
+{
+    using _ValueType = typename std::ranges::range_value_t<_R>;
+
+    if constexpr (!std::is_trivially_destructible_v<_ValueType>)
+    {
+        oneapi::dpl::__internal::__ranges::__pattern_walk_n(
+            __tag, std::forward<_ExecutionPolicy>(__exec),
+            oneapi::dpl::__internal::__op_destroy<std::decay_t<_ExecutionPolicy>>{},
+            oneapi::dpl::__ranges::views::all(std::forward<_R>(__r)));
+    }
+
+    return std::ranges::borrowed_iterator_t<_R>{std::ranges::end(__r)};
```

Comment:
Are we guaranteed that `end()` returns an iterator rather than a sentinel here?  
Should we instead use begin() + size()? 

Same questions in a few places in this file.

Reply:
Fixed the same error in other PR's too.

Reply:
@danhoeflinger, resolved?